### PR TITLE
fix(github): Massage links to issues, pulls or discussions

### LINKED ISSUE #10902: Renovate spam GitHub issues
- [x] WhiteSource Renovate hosted app on github.com
- [ ] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [x] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [ ] GitLab self-hosted

Renovate version: any

**Describe the bug**

Renovate opens pull requests linking the PRs on the changelog without any proxies. This makes the work of the maintainers harder. We can no longer easily find all the related effort to a given issue. 

**Have you created a minimal reproduction repository?**

See https://github.com/mui-org/material-ui/issues/27170

Capture d’écran 2021-07-20 à 12 34 21

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

While replacing for all GitHub links turned to be painful, the more restricted subset (discussions, issues and pulls) should protect us from spamming mentioned repositories.

## Context:

Closes #10902
Created instead of #11133

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by HonkingGoose** [comment]
Seems there's a problem with the Jest snapshots?

**Comment by zharinov** [comment]
Ah, thanks a lot. I've enfored preserving of whilte spaces code, which broke old snapshot.

**Comment by HonkingGoose** [comment]
> Ah, thanks a lot. I've enfored preserving of whilte spaces code, which broke old snapshot.

You're welcome. By the way you also missed clicking the checkboxes in the PR body text. 😉

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
What about users? if there's a link to `https://github.com/rarkins` in a PR body, will I get notified?

File: lib/platform/github/massage-markdown-links.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+const urlRegex =
+  /(?:https?:)?(?:\/\/)?(?:www\.)?(?<!api\.)(?:to)?github\.com\/[-_a-z0-9]+\/[-_a-z0-9]+\/(?:discussions|issues|pull)\/[0-9]+(?:#[-_a-z0-9]+)?/i;
+
+function massageLink(input: string): string {
+  return input.replace(/(?:to)?github\.com/i, 'togithub.com');
+}
+
+function collectLinkPosition(input: string, matches: UrlMatch[]): Plugin<any> {
+  const transformer = (tree: any): void => {
```
Comment (viceice) [inline lib/platform/github/massage-markdown-links.ts:20]: Dont' we have types for tree? 😕

**Review by viceice — COMMENTED** [review]
Beside types LGTM

File: lib/platform/github/massage-markdown-links.ts (hunk 1 above)
Comment (zharinov) [reply lib/platform/github/massage-markdown-links.ts:20]: Seems like it's a pretty big mismatch between types and real objects I saw in debugger

Comment (viceice) [reply lib/platform/github/massage-markdown-links.ts:20]: Can we add our own minimal tree interface then?

File: lib/platform/github/massage-markdown-links.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import remark from 'remark';
+import type { Plugin } from 'unified';
+import { logger } from '../../logger';
+
+interface UrlMatch {
+  start: number;
+  end: number;
+  replaceTo: string;
+}
+
```
Comment (viceice) [inline lib/platform/github/massage-markdown-links.ts:11]:
```suggestion

type TreeNode = Parent | Link | Text;
```

File: lib/platform/github/massage-markdown-links.ts (hunk 1 above)
Comment (viceice) [reply lib/platform/github/massage-markdown-links.ts:20]:
```suggestion
  const transformer = (tree: TreeNode): void => {
```

Can this work?

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**